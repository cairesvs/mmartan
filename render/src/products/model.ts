import { ProductCategory } from "./category";

export class Product {
    constructor(readonly name: string, readonly description: string, readonly photos: string[],
        readonly price: number, readonly discount: number, readonly category: ProductCategory, readonly id?: number) { }
}