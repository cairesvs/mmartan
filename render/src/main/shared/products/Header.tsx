import * as React from 'react';
const logo = require('./logo.svg') as string;
const search = require('./search.svg') as string;


export interface HeaderProps {
    submitFn: (event: React.FormEvent<HTMLFormElement>) => void;
    changeTermFn: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

export class Header extends React.Component<HeaderProps, {}> {
    constructor(props: HeaderProps) {
        super(props);
    }

    render() {
        return (
            <header className='default-header'>
                <img src={logo} />
                <div className='search-container'>
                    <form onSubmit={this.props.submitFn}>
                        <input type='search' className='search' placeholder='Buscar produtos...' onChange={this.props.changeTermFn} />
                    </form>
                </div>
            </header>
        );
    }
}


