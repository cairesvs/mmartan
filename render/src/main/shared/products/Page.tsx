import * as React from 'react';
import { Link } from 'react-router-dom';

export interface PageProps {
    changePageSizeFn: (event: React.ChangeEvent<HTMLSelectElement>) => void;
    changePageFn: (page: number) => (event: React.MouseEvent<HTMLAnchorElement>) => void;
    pages: number[];
    term: string;
    pageSize: number;
}

export class Page extends React.Component<PageProps, {}> {
    constructor(props: PageProps) {
        super(props);
    }

    render() {
        return (
            <div className='products-page'>
                <div className='products-page-size'>
                    <select onChange={this.props.changePageSizeFn}>
                        <option value='16'>16 produtos por página</option>
                        <option value='30'>30 produtos por página</option>
                        <option value='50'>50 produtos por página</option>
                    </select>
                </div>
                <div className='products-page-number'>
                    {this.props.pages &&
                        this.props.pages.map((page, i) =>
                            <Link
                                key={i}
                                onClick={this.props.changePageFn(page)}
                                to={`/?q=${this.props.term}&size=${this.props.pageSize}&page=${page}`}>{page}</Link>
                        )}
                </div>
            </div>
        );
    }
}


