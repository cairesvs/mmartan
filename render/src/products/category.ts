export enum ProductCategory {
    SHEETS = "Lençol",
    BED = "Cama",
    COUCH = "Sofá"
}