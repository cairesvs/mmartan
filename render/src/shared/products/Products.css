.productslist {
  margin: 50px auto;
  width: 1024px;
}

.products-page {
  margin: 50px auto;
  width: 1024px;
  display: flex;
  justify-content: space-between;
}

.products-page select {
  width: 200px;
}

.products-page-number a {
  padding: 5px;
  text-decoration: none;
  color: black;
  margin: 10px;
}

.productslist small {
  color: #828282;
}

.header {
  color: #666666;
  padding: 3px;
  height: 28px;
  border-bottom: 2px solid #dfbe7f;
  margin-bottom: 40px;
}

.header-title {
  float: left;
}

.sort {
  float: right;
  font-size: 13px;
  color: #c6b0f1;
  margin: 4px 15px 0 0;
}

.sort a {
  text-decoration: none;
  margin: 0 5px;
  color: #c6b0f1;
}

.sort-item {
  cursor: pointer;
}
.sort-selected {
  font-weight: bold;
}
.logo {
  margin-top: 2px;
  margin-right: 10px;
  border: 1px white solid;
  vertical-align: bottom;
}

.products-item {
  padding: 0 15px;
}

.products-position {
  color: #828282;
}

.products-details {
  display: block;
  margin-left: 40px;
  margin-top: -12px;
}

.products {
    font-family: "Avenir Light", Helvetica, Arial, sans-serif;
  }
  
  .products::after {
    display: block;
    content: "";
    clear: both;
  }
  
  .selected {
    font-weight: bold;
  }
  
  header {
    background-color: #ecebf0;
    padding: 20px;
    display: flex;
    border-bottom: 1px solid lightgray;
  }

  .search-container {
    margin-left: auto;
  }

  .search-container form {
    width: 400px;
  }
  
  .search {
    padding:6px 15px 6px 30px;
    margin:3px;
    background-image: url(/media/search.svg);
    background-repeat: no-repeat;
    background-size: contain;
    border-radius:15px; 
    width: 100%;
  }

  .typed {
    background-color: #ecebf0;
    color: #433e66;
    font-size: 2em;
    padding: 40px 20px;
    border-bottom: 1px solid lightgray;
  }
  
  aside {
    display: block;
    position: absolute;
    text-align: right;
    width: 8%;
    margin-left: -6%;
    color: #999;
  }
  
  article {
    float: left;
    padding: 0 0 0 52px;
    width: 35%;
  }
  
  figure {
    float: right;
    width: 48.5%;
  }

  .products-info {
    display: flex;
    justify-content: space-between;
  }

  .products-info-detail {
    width: 50%;
  }

  .products-info-price {
    width: 20%;
  }

  .product-has-discount {
    text-decoration: line-through;
  }
