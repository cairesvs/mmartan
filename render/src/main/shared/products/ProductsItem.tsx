import * as React from 'react';
import { Product } from '../../products/model';

export class ProductsItem extends React.Component<{ product: Product }, any> {
    render() {
        const product = this.props.product;
        const hasDiscount = product.discount > 0;
        let finalPrice = product.price;
        if (hasDiscount) finalPrice = product.price - (product.price * (product.discount / 100));
        return (
            <div className='products-info'>
                <div className='products-info-images'>
                    {product.photos.slice(0, 4).map((photo, i) => <img width='80' height='80' key={i} src={photo} />)}
                </div>
                <div className='products-info-detail'>
                    <p>{product.name}</p>
                    <p>{product.description}</p>
                </div>
                {!hasDiscount &&
                    <div className='products-info-price'>
                        <span>R${product.price}</span>
                    </div>
                }
                {hasDiscount &&
                    <div className='products-info-price'>
                        <span className='product-has-discount'>R${product.price}</span> por <span>R${finalPrice}</span>
                    </div>
                }
            </div>
        );
    }
}
