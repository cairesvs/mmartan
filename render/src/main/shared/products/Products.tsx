import * as React from 'react';
import './Products.css';
import * as fetch from 'isomorphic-fetch';
import { Product } from '../../products/model';
import { ProductsList, ProductsListProps } from './ProductsList';
import { RouteComponentProps, Link } from 'react-router-dom';
import { Header } from './Header';
import { Typed } from './Typed';
import { Page } from './Page';

declare let BACKEND_HOST: string;
const host = BACKEND_HOST;

export interface ProductsProps extends RouteComponentProps<ProductsProps> {
  data: Product[];
  term: string;
  total: number;
  pageSize: number;
  page: number;
}

export class Products extends React.Component<ProductsProps, ProductsProps> {
  constructor(props: ProductsProps) {
    super(props);
    let data: Product[];
    let term: string;
    let total: number;
    let pageSize: number = 16;
    let page: number = 1;
    if (props.staticContext) {
      const sc = props.staticContext;
      data = !sc.results.error && sc.results.results || [];
      total = !sc.results.error && sc.results.total.count || 0;
      term = props.staticContext.term || '';
      pageSize = props.staticContext.pageSize || 16;
      page = props.staticContext.page || 1;
    } else {
      const dataWindow = (window as any).__data__;
      const termWindow = (window as any).__term__;
      data = !dataWindow.error && dataWindow.results || [];
      total = !dataWindow.error && dataWindow.total.count || 0;
      term = termWindow || '';
    }

    this.state = {
      total: total,
      term: term,
      data: data,
      page: page,
      pageSize: pageSize
    } as ProductsProps;

    this.changeTerm = this.changeTerm.bind(this);
    this.submit = this.submit.bind(this);
    this.changePageSize = this.changePageSize.bind(this);
    this.changePage = this.changePage.bind(this);
  }

  changePage(page: number) {
    return (event: React.MouseEvent<HTMLAnchorElement>) => {
      this.props.history.push(`?q=${this.state.term}&size=${this.state.pageSize}&page=${page}`);
      event.preventDefault();
      Products.getInitialData(this.state.term, this.state.pageSize, page)
        .then(data => {
          if (!data.error)
            this.setState({ data: data.results, total: data.total.count, page: page });
        })
        .catch(error => console.log(error));
    };
  }

  changePageSize(event: React.ChangeEvent<HTMLSelectElement>) {
    const pageSize = parseInt(event.target.value, 10);
    this.props.history.push(`?q=${this.state.term}&size=${pageSize}&page=${this.state.page}`);
    Products.getInitialData(this.state.term, pageSize, this.state.page)
      .then(data => {
        if (!data.error)
          this.setState({ data: data.results, total: data.total.count, pageSize: pageSize });
      })
      .catch(error => console.log(error));
  }

  changeTerm(event: React.ChangeEvent<HTMLInputElement>) {
    this.setState({ page: 1, term: event.target.value });
  }

  submit(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault();
    this.props.history.push(`?q=${this.state.term}&size=${this.state.pageSize}&page=${this.state.page}`);
    Products.getInitialData(this.state.term, this.state.pageSize, this.state.page)
      .then(data => {
        if (!data.error)
          this.setState({ data: data.results, total: data.total.count });
      })
      .catch(error => console.log(error));
  }

  render() {
    const products = this.state.data;
    const total = this.state.total;
    const maxPage = Math.floor(total / this.state.pageSize);
    const pages = [...Array(maxPage).keys()].map((n) => n + 1);
    return (
      <div className='products'>
        <Header submitFn={this.submit} changeTermFn={this.changeTerm} />
        <Typed term={this.state.term} />
        <ProductsList total={total} products={products} />
        <Page changePageSizeFn={this.changePageSize} changePageFn={this.changePage}
          pages={pages} term={this.state.term} pageSize={this.state.pageSize} />
      </div>
    );
  }

  static getInitialData(term: string, size: number = 16, page: number = 1): Promise<any> {
    return fetch(`${host}/products/search?q=${term}&size=${size}&page=${page}`)
      .then(response => response.json());
  }
}

export default Products;
