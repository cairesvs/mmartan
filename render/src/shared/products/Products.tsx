import * as React from 'react';
import './Products.css';
import * as fetch from 'isomorphic-fetch';
import { Product } from '../../products/model';
import { ProductsList, ProductsListProps } from './ProductsList';
import { RouteComponentProps } from 'react-router-dom';

declare let BACKEND_HOST: string;
const host = BACKEND_HOST;

export interface ProductsProperties extends RouteComponentProps<ProductsProperties> { data: Product[]; term: string; }

export class Products extends React.Component<ProductsProperties, ProductsListProps> {
  constructor(props: ProductsProperties) {
    super(props);
    let data: Product[];
    if (props.staticContext) {
      data = props.staticContext.results;
    } else {
      data = (window as any).__data__;
      delete (window as any).__data__;
    }
    this.state = { products: data };
  }

  render() {
    const products = this.state.products;
    return <ProductsList products={products} />;
  }

  static getResponse(term: string): Promise<any> {
    return fetch(`${host}/products/search?q=${term}`)
      .then(response => response.json());
  }
}

export default Products;
