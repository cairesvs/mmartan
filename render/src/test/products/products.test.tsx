import * as React from 'react';
import * as ReactDOM from 'react-dom';
import * as TestUtils from 'react-dom/test-utils';
import * as renderer from 'react-test-renderer';
import { Products, ProductsProps } from '../../main/shared/products/Products';
import { RouteComponentProps } from 'react-router';
import { UnregisterCallback, Href } from 'history';
import { Header } from '../../main/shared/products/Header';
import { Page } from '../../main/shared/products/Page';
import { Product } from '../../main/products/model';
import { ProductsItem } from '../../main/shared/products/ProductsItem';
import { ProductCategory } from '../../main/products/category';
import { ProductsList } from '../../main/shared/products/ProductsList';

it('<Header/>', () => {
    const data = {
        results: [],
        total: {
            count: 0
        }
    };
    const term: string = '';
    const total: number = 0;
    const pageSize: number = 16;
    const page: number = 1;
    const submitFn = (event: any) => { };
    const changeTermFn = (event: any) => { };
    const component = renderer.create(
        <Header submitFn={submitFn} changeTermFn={changeTermFn} />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

it('<Page/>', () => {
    const changePageFn = (page: number) => (event: React.MouseEvent<HTMLAnchorElement>) => { };
    const changePageSizeFn = (event: React.ChangeEvent<HTMLSelectElement>) => { };
    const pages = [] as number[];
    const term = '';
    const pageSize = 1;
    const component = renderer.create(
        <Page changePageFn={changePageFn} changePageSizeFn={changePageSizeFn} pages={pages} term={term} pageSize={pageSize} />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

it('<ProductsList/>', () => {
    const product = new Product('Sofá', 'Lindo sofá', [], 1000, 0, ProductCategory.COUCH, 1);
    const props = { total: 1, products: [product] };
    const component = renderer.create(
        <ProductsList products={props.products} total={props.total} />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

it('<ProductsItem/>', () => {
    const product = new Product('Sofá', 'Lindo sofá', [], 1000, 0, ProductCategory.COUCH, 1);
    const component = renderer.create(
        <ProductsItem product={product} />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

it('<ProductsItem/> without discount', () => {
    const product = new Product('Sofá', 'Lindo sofá', ['http://example.com/image.png'], 1000, 0, ProductCategory.COUCH, 1);
    const productItem: any = TestUtils.renderIntoDocument(
        <ProductsItem product={product} />
    );

    const productItemNode = ReactDOM.findDOMNode(productItem);
    const priceInfo = productItemNode.getElementsByClassName('products-info-price');
    expect(priceInfo.length).toEqual(1);

    expect(priceInfo[0].textContent).toEqual('R$1000');
});

it('<ProductsItem/> with discount', () => {
    const product = new Product('Sofá', 'Lindo sofá', ['http://example.com/image.png'], 1000, 10, ProductCategory.COUCH, 1);
    const productItem: any = TestUtils.renderIntoDocument(
        <ProductsItem product={product} />
    );

    const productItemNode = ReactDOM.findDOMNode(productItem);
    const priceInfo = productItemNode.getElementsByClassName('products-info-price');
    expect(priceInfo.length).toEqual(1);

    expect(priceInfo[0].textContent).toEqual('R$1000 por R$900');
});
