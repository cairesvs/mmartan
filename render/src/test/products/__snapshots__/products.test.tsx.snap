// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header/> 1`] = `
<header
  className="default-header"
>
  <div
    className="logo"
  />
  <div
    className="search-container"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        className="search"
        onChange={[Function]}
        placeholder="Buscar produtos..."
        type="search"
      />
    </form>
  </div>
</header>
`;

exports[`<Page/> 1`] = `
<div
  className="products-page"
>
  <div
    className="products-page-size"
  >
    <select
      onChange={[Function]}
    >
      <option
        value="16"
      >
        16 produtos por página
      </option>
      <option
        value="30"
      >
        30 produtos por página
      </option>
      <option
        value="50"
      >
        50 produtos por página
      </option>
    </select>
  </div>
  <div
    className="products-page-number"
  />
</div>
`;

exports[`<ProductsItem/> 1`] = `
<div
  className="products-info"
>
  <div
    className="products-info-images"
  />
  <div
    className="products-info-detail"
  >
    <p>
      Sofá
    </p>
    <p>
      Lindo sofá
    </p>
  </div>
  <div
    className="products-info-price"
  >
    <span>
      R$
      1000
    </span>
  </div>
</div>
`;

exports[`<ProductsList/> 1`] = `
<div
  className="productslist"
>
  <div
    className="header"
  >
    <div
      className="header-title"
    >
      <strong>
        1
         produtos encontrados
      </strong>
    </div>
  </div>
  <div
    className="products-info"
  >
    <div
      className="products-info-images"
    />
    <div
      className="products-info-detail"
    >
      <p>
        Sofá
      </p>
      <p>
        Lindo sofá
      </p>
    </div>
    <div
      className="products-info-price"
    >
      <span>
        R$
        1000
      </span>
    </div>
  </div>
</div>
`;
